File: server/src/main/java/com/barataribeiro/taskr/notification/Notification.java
package com.barataribeiro.taskr.notification;

import com.barataribeiro.taskr.user.User;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.io.Serial;
import java.io.Serializable;
import java.time.Instant;

@Builder
@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@ToString
@Entity
@Table(name = "tb_notifications", indexes = {
        @Index(name = "idx_notification_id_unq", columnList = "id, recipient_id", unique = true)
})
public class Notification implements Serializable {
    @Serial
    private static final long serialVersionUID = 1L;

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(updatable = false, nullable = false, unique = true)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String message;

    @Builder.Default
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private NotificationType type = NotificationType.INFO;

    @ToString.Exclude
    @ManyToOne(optional = false)
    @JoinColumn(name = "recipient_id", nullable = false)
    private User recipient;

    @Builder.Default
    @Column(name = "is_read", columnDefinition = "BOOLEAN default 'false'", nullable = false)
    private boolean isRead = false;

    @Column(updatable = false)
    @CreationTimestamp
    private Instant createdAt;

    @UpdateTimestamp
    private Instant updatedAt;
}

File: client/src/@types/notification.ts
import { Author } from "@/@types/user"

enum NotificationType {
    INFO = "INFO",
    SUCCESS = "SUCCESS",
    WARNING = "WARNING",
    ERROR = "ERROR",
}

interface Notification {
    id: number
    title: string
    message: string
    type: NotificationType
    recipient: Author
    isRead: boolean
    createdAt: string
    updatedAt: string
}

interface LatestNotifications {
    latestNotifications: Notification[]
    totalCount: number
    totalRead: number
    totalUnread: number
}

export { NotificationType }

export type { Notification, LatestNotifications }
